File: DSA/src/Questions/Graph/TripartiteGraph.java
package Questions.Graph;

public class TripartiteGraph {

    // Depth First Search (DFS) function to check if the graph is tripartite
    public static boolean dfs(int node, int c, int[] color, int[][] graph) {
        // If the color of the current node is not 0 (unvisited), check if it matches
        // the expected color 'c'
        if (color[node] != 0) {
            return color[node] == c;
        }
        // Color the current node with the given color 'c'
        color[node] = c;
        // Traverse all neighbors of the current node
        for (int neighbor : graph[node]) {
            // Recursively check if the neighbors can be colored with the opposite color
            // (-c)
            if (!dfs(neighbor, -c, color, graph)) {
                return false; // If not, return false
            }
        }
        return true; // If all neighbors can be colored properly, return true
    }

    // Function to check if the graph is tripartite
    public static boolean isTripartite(int n, int[][] graph) {
        int[] color = new int[n]; // Array to store the color of each node (0 for unvisited)
        // Traverse all nodes in the graph
        for (int i = 0; i < n; i++) {
            // If the current node is unvisited and cannot be colored properly, return false
            if (color[i] == 0 && !dfs(i, 1, color, graph)) {
                return false;
            }
        }
        return true; // If all nodes can be colored properly, return true
    }

    public static void main(String[] args) {
        int n = 4; // Number of nodes in the graph
        int[][] graph = { { 0, 1 }, { 1, 1 }, { 2, 0 } }; // Adjacency list representation of the graph
        int[][] graph1 = { { 0, 1 }, { 1, 2 }, { 2, 3 }, { 3, 0 } }; // Adjacency list representation of the graph
        System.out.println("Is the graph tripartite? " + isTripartite(n, graph1));
        System.out.println("Is the graph tripartite? " + isTripartite(n, graph));
    }
}

File: DSA/Python/Questions/Graph/tripartite_graph.py
"""
# Question: Check if a Graph is Tripartite
# Link: (Assumed hypothetical problem)

# Problem Statement:
# Given an undirected graph, determine if it is tripartite. A graph is tripartite if its vertices can be divided into three disjoint sets such that no two graph vertices within the same set are adjacent.

# Example:
# Input: n = 4, graph = [[0, 1], [1, 2], [2, 3], [3, 0]]
# Output: False
"""

from typing import List
from collections import deque

# BFS Solution
"""
Algorithm:
1. Initialize a color array to store colors of nodes.
2. Use BFS to traverse the graph.
3. Color the starting node with one color and alternate colors for adjacent nodes.
4. If a conflict is found (two adjacent nodes have the same color), return False.
5. If no conflicts are found, return True.

Time Complexity: O(V + E) where V is the number of vertices and E is the number of edges.
Space Complexity: O(V) for the color array and queue.
"""


def is_tripartite(n: int, graph: List[List[int]]) -> bool:
    color = [0] * n  # 0 means uncolored, 1 and -1 are two different colors

    def bfs(start: int) -> bool:
        queue = deque([start])
        color[start] = 1  # Start coloring with 1

        while queue:
            node = queue.popleft()
            for neighbor in graph[node]:
                if color[neighbor] == 0:  # If not colored, color with opposite color
                    color[neighbor] = -color[node]
                    queue.append(neighbor)
                elif color[neighbor] == color[node]:  # If same color, not bipartite
                    return False
        return True

    for i in range(n):
        if color[i] == 0:  # Not yet colored
            if not bfs(i):
                return False
    return True


def main():
    test_cases = [
        {"n": 4, "graph": [[1, 3], [0, 2], [1, 3], [0, 2]], "expected": True},
        {"n": 4, "graph": [[0, 1], [1, 2], [2, 3], [3, 0]], "expected": False},
        {"n": 3, "graph": [[0, 1], [1, 2], [2, 0]], "expected": False},
    ]

    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"n: {test['n']}, graph: {test['graph']}")
        result = is_tripartite(test["n"], test["graph"])
        print(f"Result: {result}, Expected: {test['expected']}")
        assert result == test["expected"], f"Test case {i + 1} failed"


if __name__ == "__main__":
    main()
